src/Integration.Vsix/SonarLintTagger/TextBufferIssueTracker.cs

```diff
@@ -0,0 +1,261 @@
+                Provider.RemoveIssueTracker(this);
+            }
+        }
+
+        private void FileActionOccurred(object sender, TextDocumentFileActionEventArgs e)
+        {
+            if (e.FileActionType == FileActionTypes.DocumentRenamed)
+            {
+                FilePath = e.FilePath;
+                ProjectItem = dte.Solution.FindProjectItem(this.FilePath);
+            }
+            else if (e.FileActionType == FileActionTypes.ContentSavedToDisk)
+            {
+                RequestAnalysis();
+            }
+        }
+
+        private void OnBufferChange(object sender, TextContentChangedEventArgs e)
+        {
+            UpdateDirtySpans(e);
+
+            var newMarkers = TranslateMarkerSpans();
+
+            SnapToNewSnapshot(newMarkers);
+        }
+
+        private void UpdateDirtySpans(TextContentChangedEventArgs e)
+        {
+            currentSnapshot = e.After;
+
+            var newDirtySpans = dirtySpans.CloneAndTrackTo(e.After, SpanTrackingMode.EdgeInclusive);
+
+            foreach (var change in e.Changes)
+            {
+                newDirtySpans = NormalizedSnapshotSpanCollection.Union(newDirtySpans, new NormalizedSnapshotSpanCollection(e.After, change.NewSpan));
+            }
+
+            dirtySpans = newDirtySpans;
+        }
+
+        private IssuesSnapshot TranslateMarkerSpans()
+        {
+            var oldSnapshot = this.Factory.CurrentSnapshot;
+            var newMarkers = oldSnapshot.IssueMarkers
+                .Select(marker => marker.CloneAndTranslateTo(currentSnapshot))
+                .Where(clone => clone != null);
+
+            return new IssuesSnapshot(this.ProjectItem.ContainingProject.Name, this.FilePath, oldSnapshot.VersionNumber + 1, newMarkers);
+        }
+
+        private bool IsValidIssueTextRange(Issue issue) =>
+            1 <= issue.StartLine && issue.EndLine <= currentSnapshot.LineCount;
+
+        private IssueMarker CreateIssueMarker(Issue issue) =>
+            issueConverter.ToMarker(issue, currentSnapshot);
+
+        private void SnapToNewSnapshot(IssuesSnapshot newIssues)
+        {
+            // Tell our factory to snap to a new snapshot.
+            this.Factory.UpdateSnapshot(newIssues);
+
+            // Tell the provider to mark all the sinks dirty (so, as a side-effect, they will start an update pass that will get the new snapshot
+            // from the factory).
+            Provider.UpdateAllSinks();
+
+            // Work out which part of the document has been affected by the changes, and tell
+            // the taggers about the changes
+            SnapshotSpan? affectedSpan = CalculateAffectedSpan(LastIssues, newIssues);
+            foreach (var tagger in activeTaggers)
+            {
+                tagger.UpdateMarkers(newIssues, affectedSpan);
+            }
+
+            this.LastIssues = newIssues;
+        }
+
+        private SnapshotSpan? CalculateAffectedSpan(IssuesSnapshot oldIssues, IssuesSnapshot newIssues)
```

Comment:
This is about the only major difference between this implementation and the VS SDK sample.
The SDK sample repeated this method in each lightweight _tagger_ class, but the calculation is the same in each case so I moved it to this class so it's only calculated once.